src/cmdstan/command_helper.hpp

```diff
@@ -0,0 +1,587 @@
+      if (stream.rdstate() & std::ifstream::failbit) {
+        std::string file_name_err
+            = std::string("\"" + file_1 + "\" and base file \"" + file + "\"");
+        std::stringstream msg;
+        msg << "Searching for  \"" << file_name_err << std::endl;
+        msg << "Can't open either of specified files," << file_name_err
+            << std::endl;
+        throw std::invalid_argument(msg.str());
+      } else {
+        return context_vector(num_chains,
+                              make_context(file, stream, file_ending));
+      }
+    } else {
+      // If we found file_1 then we'll assume file_{1...N} exists
+      context_vector ret;
+      ret.reserve(num_chains);
+      ret.push_back(make_context(file_1, stream_1, file_ending));
+      for (size_t i = 1; i < num_chains; ++i) {
+        std::string file_i
+            = std::string(file_name + "_" + std::to_string(i) + file_ending);
+        std::fstream stream_i(file_1.c_str(), std::fstream::in);
+        // If any stream fails at this point something went wrong with file
+        // names.
+        if (stream_i.rdstate() & std::ifstream::failbit) {
+          std::string file_name_err = std::string(
+              "\"" + file_1 + "\" but cannot open \"" + file_i + "\"");
+          std::stringstream msg;
+          msg << "Found " << file_name_err << std::endl;
+          throw std::invalid_argument(msg.str());
+        }
+        ret.push_back(make_context(file_i, stream_i, file_ending));
+      }
+      return ret;
+    }
+  }
+  // This should not happen
+  using stan::io::dump;
+  std::fstream stream(file.c_str(), std::fstream::in);
+  return context_vector(num_chains, std::make_shared<dump>(dump(stream)));
+}
+
+/**
+ * Get vector of constrainted parameter names from model.
+ * Throws exception when model has no parameters.
+ *
+ * @param model the instantiated model
+ * @param vector of constrained param names
+ */
+void get_constrained_params(const stan::model::model_base &model,
+                            std::vector<std::string> &param_names) {
+  model.constrained_param_names(param_names, false, false);
+  if (param_names.size() < 1) {
+    std::stringstream msg;
+    msg << "Model " << model.model_name()
+        << " has no parameters, nothing to estimate." << std::endl;
+    throw std::invalid_argument(msg.str());
+  }
+}
+
+/**
+ * Opens input stream for file.
+ * Throws exception if stream cannot be opened.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @return input stream
+ */
+std::ifstream safe_open(const std::string fname) {
+  std::ifstream stream(fname.c_str());
+  if (fname != "" && (stream.rdstate() & std::ifstream::failbit)) {
+    std::stringstream msg;
+    msg << "Can't open specified file, \"" << fname << "\"" << std::endl;
+    throw std::invalid_argument(msg.str());
+  }
+  return stream;
+}
+
+/**
+ * Parse a StanCSV output file created by the NUTS-HMC sampler and
```

Comment:
I believe this comment is incorrect, there's no restriction making this specific to NUTS (I think anyway, this seems to be based on the code changed in https://github.com/stan-dev/cmdstan/pull/1106)

Reply:
It’s used mainly by `generate_quantities`, which supports non-NUTS inputs, so this is a bit confusing

Reply:
and it works?  that's interesting.  I'll check the "stan::io::read_csv" method - maybe that code is flexible enough to handle output files from other methods as well.   I'm not fond of that dependency, FWIW.

Reply:
Yeah, I made the (minimal) change in #1106 to support them. The only thing that it directly depended on was that HMC outputs 7 things before the user parameters, so now it just counts them.

Reply:
cool!